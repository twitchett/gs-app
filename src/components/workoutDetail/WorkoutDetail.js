import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { StyleSheet, FlatList, Text, TextInput, View, Button } from 'react-native'
import { addParticipant, changeView } from '../../redux/actions'

export class WorkoutDetail extends React.Component {
  constructor(props) {
    super(props)
    this.renderItem = this.renderItem.bind(this)
    this.handleInputChange = this.handleInputChange.bind(this)
    this.handleAddButtonPress = this.handleAddButtonPress.bind(this)
    this.handleBackButtonPress = this.handleBackButtonPress.bind(this)

    this.state = {
      name: ''
    }
  }

  handleInputChange(input) {
    this.setState({ name: input })
  }

  handleAddButtonPress() {
    const { workoutId, addParticipant } = this.props
    const { name } = this.state

    addParticipant({ workoutId, name })
  }

  handleBackButtonPress() {
    const { changeViewToList } = this.props

    changeViewToList()
  }

  renderItem({ item }) {
    return (
      <Text style={styles.item}>{item}</Text>
    )
  }

  render() {
    const { title, participants } = this.props

    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 25 }}>{title}</Text>
        <View style={styles.addParticipantContainer}>
          <TextInput
            style={styles.addInput}
            placeholder="Add participant"
            onChangeText={this.handleInputChange}
          />
          <Button
            title="Add"
            onPress={this.handleAddButtonPress}
          />
        </View>
        <FlatList
          data={participants}
          keyExtractor={(item, index) => index.toString()}
          renderItem={this.renderItem}
        />
        <Button
          title="Back"
          onPress={this.handleBackButtonPress}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    paddingTop: 22,
    justifyContent: 'center',
  },
  addParticipantContainer: {
    flexDirection: 'row',
    padding: 20
  },
  item: {
    padding: 10,
    fontSize: 18,
    height: 44,
  },
  addInput: {
    height: 40,
    width: 150
  }
})

WorkoutDetail.propTypes = {
  workoutId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  addParticipant: PropTypes.func.isRequired,
  changeViewToList: PropTypes.func.isRequired,
  participants: PropTypes.arrayOf(PropTypes.string),
}

WorkoutDetail.defaultProps = {
  participants: []
}

const mapStateToProps = state => {
  const workoutId = state.view
  const { title, participants } = state.workouts[workoutId]
  return {
    workoutId,
    title,
    participants
  }
}

const mapDispatchToProps = dispatch => ({
  addParticipant: payload => { dispatch(addParticipant(payload)) },
  changeViewToList: () => { dispatch(changeView('workoutList')) }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WorkoutDetail)