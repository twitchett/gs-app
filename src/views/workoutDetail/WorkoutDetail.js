import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { StyleSheet, FlatList, Text, TextInput, View, Button } from 'react-native'
import { addParticipant } from '../../reducer'

export class WorkoutDetail extends React.Component {
  constructor(props) {
    super(props)
    this.renderItem = this.renderItem.bind(this)
    this.handleOnInputChange = this.handleOnInputChange.bind(this)
    this.handleOnButtonPress = this.handleOnButtonPress.bind(this)

    this.state = {
      newParticipant: ''
    }
  }

  handleOnInputChange(input) {
    this.setState({ newParticipant: input })
  }

  handleOnButtonPress() {
    const { workoutId, addParticipant } = this.props
    const { newParticipant } = this.state

    addParticipant({ workoutId, participant: newParticipant })
  }

  renderItem({ item }) {
    return (
      <Text style={styles.item}>{item}</Text>
    )
  }

  render() {
    const { title, participants } = this.props
    const { newParticipant } = this.state

    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 25 }}>{title}</Text>
        <View style={styles.addParticipantContainer}>
          <TextInput
            style={styles.addInput}
            placeholder="Add participant"
            onChangeText={this.handleOnInputChange}
          />
          <Button
            title="Add"
            onPress={this.handleOnButtonPress}
          />
        </View>
        <FlatList
          data={participants}
          keyExtractor={(item, index) => index.toString()}
          renderItem={this.renderItem}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    paddingTop: 22,
    justifyContent: 'center',
  },
  addParticipantContainer: {
    flexDirection: 'row',
    padding: 20
  },
  item: {
    padding: 10,
    fontSize: 18,
    height: 44,
  },
  addInput: {
    height: 40,
    width: 150
  }
})

WorkoutDetail.propTypes = {
  workoutId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  addParticipant: PropTypes.func.isRequired,
  participants: PropTypes.arrayOf(PropTypes.string),
}

WorkoutDetail.defaultProps = {
  participants: []
}

const mapStateToProps = state => {
  const workoutId = state.view
  const { title, participants } = state.workouts[workoutId]
  return {
    workoutId,
    title,
    participants
  }
}

const mapDispatchToProps = dispatch => ({
  addParticipant: payload => { dispatch(addParticipant(payload)) }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WorkoutDetail)