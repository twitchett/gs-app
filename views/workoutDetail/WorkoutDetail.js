import React from 'react'
import { StyleSheet, FlatList, Text, TextInput, View, Button } from 'react-native'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

export class WorkoutDetail extends React.Component {
  constructor(props) {
    super(props)
    this.renderItem = this.renderItem.bind(this)
  }

  renderItem({ item }) {
    return (
      <Text style={styles.item}>{item}</Text>
    )
  }

  render() {
    const { title, participants } = this.props

    return (
      <View style={styles.container}>
        <Text style={{ fontSize: 25 }}>{title}</Text>
        <View style={styles.addParticipantContainer}>
          <TextInput
            style={styles.addInput}
            placeholder="Add participant"
            onChangeText={text => this.setState({text})}
          />
          <Button
            title="Add"
            onPress={() => { /* TODO */ }}
          />
        </View>
        <FlatList
          data={participants}
          keyExtractor={(item, index) => index.toString()}
          renderItem={this.renderItem}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    paddingTop: 22,
    justifyContent: 'center',
  },
  addParticipantContainer: {
    flexDirection: 'row',
    padding: 20
  },
  item: {
    padding: 10,
    fontSize: 18,
    height: 44,
  },
  addInput: {
    height: 40,
    width: 150
  }
})

WorkoutDetail.propTypes = {
  workoutId: PropTypes.string.isRequired,
  participants: PropTypes.arrayOf(PropTypes.string)
}

WorkoutDetail.defaultProps = {
  participants: []
}

const mapStateToProps = (state, ownProps) => {
  const { title, participants } = state.workouts[ownProps.workoutId]
  return {
    title,
    participants
  }
}

const mapDispatchToProps = {}

export default connect(mapStateToProps, mapDispatchToProps)(WorkoutDetail);