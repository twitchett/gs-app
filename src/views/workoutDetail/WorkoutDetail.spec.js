import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { WorkoutDetail } from './WorkoutDetail'

describe('WorkoutDetail', () => {
  const defaultProps = {
    workoutId: 'running',
    title: 'Running',
    addParticipant: () => {}
  }

  it('renders the component', () => {
    const tree = renderer.create(<WorkoutDetail {...defaultProps} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('displays the list of participants', () => {
    const props = {
      ...defaultProps,
      participants: ['Kim', 'Tim', 'Jim'],
    }
    const tree = renderer.create(<WorkoutDetail {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  xit('updates the state when input is entered', () => {
    const wrapper = mount(<WorkoutDetail {...defaultProps} />)
    const textInput = wrapper.find('TextInput').first()

    expect(wrapper.state('name')).to.equal('')

    // TODO: how to simulate text input?
    // textInput.simulate('change', 'The Doctor')
    wrapper.update()

    expect(wrapper.state('name')).to.equal('The Doctor')
  })

  it('calls addParticipant when the Add button is pressed', () => {
    const addParticipantMock = jest.fn()
    const props = {
      ...defaultProps,
      addParticipant: addParticipantMock
    }
    const wrapper = mount(<WorkoutDetail {...props} />)
    const textInput = wrapper.find('TextInput').first()

    // TODO: same problem as above
    // textInput.simulate('change', 'The Doctor')
    wrapper.update()

    const button = wrapper.find('Button').first().props().onPress()

    expect(addParticipantMock).toHaveBeenCalledTimes(1)
    expect(addParticipantMock).toHaveBeenCalledWith({
      name: '',
      workoutId: 'running'
    })
  })
})
