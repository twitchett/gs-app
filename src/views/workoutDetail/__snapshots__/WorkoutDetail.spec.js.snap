// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutDetail displays the list of participants 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 22,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 25,
      }
    }
  >
    Running
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "padding": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Add participant"
      style={
        Object {
          "height": 40,
          "width": 150,
        }
      }
    />
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        "Kim",
        "Tim",
        "Jim",
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 18,
              "height": 44,
              "padding": 10,
            }
          }
        >
          Kim
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 18,
              "height": 44,
              "padding": 10,
            }
          }
        >
          Tim
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "fontSize": 18,
              "height": 44,
              "padding": 10,
            }
          }
        >
          Jim
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Back
      </Text>
    </View>
  </View>
</View>
`;

exports[`WorkoutDetail renders the component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 22,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 25,
      }
    }
  >
    Running
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "padding": 20,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Add participant"
      style={
        Object {
          "height": 40,
          "width": 150,
        }
      }
    />
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={undefined}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Add
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    invertStickyHeaders={undefined}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Back
      </Text>
    </View>
  </View>
</View>
`;
